### PR TITLE
Feat / Kujira

### PR DESCRIPTION
**Pull request by danilo-silva-funttastic**
**Summary**
The Funttastic Labs and Robotter.AI team requests the inclusion of the Kujira DEX connector to the Hummingbot Gateway and Client. Kujira will enhance Humminbot features and capabilities and will be able to use any of the Hummingbot strategies and/or the customized scripts.

**PRP**: https://snapshot.org/#/hbot-prp.eth/proposal/0x7dbd4a6f3cc7460ca6f56415a57b9727ca7a9227be625efdc4e71dee3d0d0781
**Tags**: DEX, Gateway, Client, Connector, Kujira
**Github handles**: https://github.com/funttastic and https://github.com/yourtrading-ai
**Kujira Blockchain Website**: https://kujira.app/

**Motivation**
- High-Quality Code and Robust Testing Coverage:
- The Kujira connector will boast high-quality code implementation with a well-designed API and a robust testing coverage.
- Also, our API also counts with more than the common endpoints but with a great addition of other ones where you can to retrieve one, several or all (when applicable) markets, order books, tickers, and transactions at once; also create, cancel, and get one, several, and all (when applicable) orders at once, with buy or sell sides, of types market or limit, and so on.
- Our test coverage counts with unit and the integration tests in a full flow manner, where we developed an entire flow covering all the users operations, guaranteeing in this way that our users are always covered, not matter the changes.
- Opening the Door to Further Connectors: The Kujira connector, using the CLOB SPOT connector standard implemented in Injective, will pave the way for the creation of additional connectors in the future, thus enhancing the flexibility and capabilities of Hummingbot.
- Kujira Key Points:
- Sustainable Sovereign Layer 1 Ecosystem
- Focuses on sustainable FinTech solutions for protocols, builders, and web3 users
- Offers a range of innovative and user-friendly products that generate revenue and aim to transform payment infrastructure
- Provides easy-to-use, cost-effective tools for people of all levels of crypto experience
- Inclusive of all blockchains and networks, and offers products that do not require participants to risk more than they can afford to lose
- Aims to allow retail investors to access opportunities previously only available to the wealthy and elite
- Mission is to help people generate income regardless of market direction
- Semi-Permissioned:
- On-chain Scheduler:
- Native Token Generation:
- Build with Cosmos SDK:
- CosmWASM 1.0 Support:
- Thriving Community:
- Built to Last
- Kujira Innovative and User-friendly tools: Kujira's decentralized ecosystem offers a range of innovative, user-friendly tools and products. Some of they are:
- ORCA: Liquidate multiple assets across the whole of Cosmos at the click of a button. No bots, no code, no problems.
- FIN: Cosmos' first decentralized, permissionless, 100% on-chain order book style token exchange.
- BOW: BOW delivers a better trading experience on FIN by narrowing the price spread deepening the pool.
- GHOST: Mobilize your assets by lending them out to earn interest and secure working liquidity by using them as collateral.
- PILOT: Launch tokens in a fair and equitable manner that benefits and gives equal access to all.), BLUE (Your KUJI Dashboard. Stake KUJI, Vote on Governance, Swap KUJI, Bridge to Kujira Network, and claim rewards; all in one place.
- FINDER: A tool to search for present and historical data on the Kujira blockchain.

### CODE REVIEW

**Comment by nikspz** [comment]
Snapshot:
https://snapshot.org/#/hbot-ip.eth/proposal/0xf835f1a3bf15140448b201b41093d86b41763c53775d528d6537a210bc20f1c7

**Comment by nikspz** [comment]
hi @danilo-silva-funttastic 
Please be informed about hummingbot ongoing migration to python3.10 this week.
The new PR referring to the https://github.com/hummingbot/hummingbot/pull/6389 will be created and merged before all the connectors PRs to prevent connectors conflicts with python3.10 environment

**Comment by danilo-silva-funttastic** [comment]
> hi @danilo-silva-funttastic Please be informed about hummingbot ongoing migration to python3.10 this week. The new PR referring to the #6389 will be created and merged before all the connectors PRs to prevent connectors conflicts with python3.10 environment

Ok, thanks for letting me know @nikspz, we will soon formalize the PR.
We will try to make it compatible with python3.10, nice upgrade by the way.

**Comment by danilo-silva-funttastic** [comment]
> @nikspz

We have already merged with the latest changes from development, and it seems the PR you mentioned has already been merged, so we might be good to go ;)

**Comment by nikspz** [comment]
FYI: test with https://github.com/hummingbot/gateway/pull/138 PR

**Comment by danilo-silva-funttastic** [comment]
@nikspz, as I was commenting with @fengtality some time before, it seems the unit tests are not all passing in the development branch, even when using the official HB code without modifications.

In this case our build hasn't passed but because of something else, Huobi connector:

```
test_listen_for_subscriptions_raises_logs_exception (test.hummingbot.connector.exchange.huobi.test_huobi_api_order_book_data_source.HuobiAPIOrderBookDataSourceUnitTests) ... E0630 12:51:09.385424818    7430 hpack_parser.cc:833]                  Error parsing 'content-type' metadata: error=invalid value key=content-type
make: *** [Makefile:13: test] Terminated
```

We can then ignore this? How should we proceed? Should we rerun the build?
The PRP is still open for voting but I think it is good to be preparing everything here in advance :)

**Review by fengtality — CHANGES_REQUESTED** [review]
@danilo-silva-funttastic The Kujira changes should primarily be limited to what's in the Gateway PR + the Kujira data feed, similar to how the client handles Dexalot and Injective.

The HIP only grants us the bandwidth to review Kujira-related code. I understand that these other changes might be helpful, but you should put them in a separate PR and submit a PRP for them.

**Review by fengtality — COMMENTED** [review]
Also, note that after @vic-en finishes the PR that addresses this issue https://github.com/hummingbot/gateway/issues/104, this PR will likely need to be refactored

**Comment by danilo-silva-funttastic** [comment]
> @danilo-silva-funttastic The Kujira changes should primarily be limited to what's in the Gateway PR + the Kujira data feed, similar to how the client handles Dexalot and Injective.
> 
> The HIP only grants us the bandwidth to review Kujira-related code. I understand that these other changes might be helpful, but you should put them in a separate PR and submit a PRP for them.

We are working here to update the branch and remove the "extra" code.

But can we take a more simple approach?

We would like to put all of the other changes in a "feat/core-improvements", create a PR, but avoid creating a PRP and a 1 week voting process to fix or improve Hummingbot behavior, where some are:
- Improving the script capabilities to be able to have async initialization, stop and on_tick methods (where the common behavior will still be the same and the other scripts should be able to operate with the same code)
- Fixing a bug on the jest configuration that doesn't show information from nested tests
- Adding .iml to .gitignore to also ignore pycharm files
- Removing an old Serum script file

As you can see some simple improvements

**Comment by danilo-silva-funttastic** [comment]
> Also, note that after @vic-en finishes the PR that addresses this issue [hummingbot/gateway#104](https://github.com/hummingbot/gateway/issues/104), this PR will likely need to be refactored

Not sure how the referred issue will be solved, but we can update our endpoints to rely on /chain/kujira here if you want.
This might not be difficult for us. We are always requiring the chain, network, and connector parameters.

For now our endpoints are /kujria

How do you suggest us to proceed?

**Comment by fengtality** [comment]
> We would like to put all of the other changes in a "feat/core-improvements", create a PR, but avoid creating a PRP

Yes, this is fine for a bunch of small fixes - note that can review and merge bug fixes without going through the governance process. 

However, I think this one should be a separate PR as it seems larger:
* Improving the script capabilities to be able to have async initialization, stop and on_tick methods (where the common behavior will still be the same and the other scripts should be able to operate with the same code)

**Comment by fengtality** [comment]
> How do you suggest us to proceed?

Please take a look at the changes in https://github.com/hummingbot/gateway/pull/147

**Comment by danilo-silva-funttastic** [comment]
> > We would like to put all of the other changes in a "feat/core-improvements", create a PR, but avoid creating a PRP
> 
> Yes, this is fine for a bunch of small fixes - note that we can review and merge bug fixes without going through the governance process.
> 
> However, I think this one should be a separate PR as it seems larger:
> 
> * Improving the script capabilities to be able to have async initialization, stop and on_tick methods (where the common behavior will still be the same and the other scripts should be able to operate with the same code)

Ok, we will proceed this way :)

**Comment by danilo-silva-funttastic** [comment]
> > How do you suggest us to proceed?
> 
> Please take a look at the changes in [hummingbot/gateway#147](https://github.com/hummingbot/gateway/pull/147)

Ok, we will be proceeding in a similar manner that we can find there on the referred PR

**Review by danilo-silva-funttastic — COMMENTED** [review]
We have reviewed all of the changes and removed all of the "extra" code from the PR.
We have also tested our code, and unit tests.
Our coverage is now in 81% for our connector.

**Comment by danilo-silva-funttastic** [comment]
> 

@fengtality, @nikspz 

We have reviewed all of the changes and removed all of the "extra" code from the PR.
We have also tested our code, and unit tests.
Our coverage is now in 81% for our connector.
We might be ready for merging.
We kindly ask you all to review it and all being ok approve our merging.

**Review by danilo-silva-funttastic — COMMENTED** [review]
@fengtality, @nikspz , @rapcmia

We have refactored our code and we tried to make the changes to be as simple as possible for the PR.

Please kindly review when you have time.

**Comment by cardosofede** [comment]
@danilo-silva-funttastic test are failing

**Comment by danilo-silva-funttastic** [comment]
> @danilo-silva-funttastic test are failing

Hey @cardosofede, we've done some new changes to the PR and I think the test is now passing. Can you please run the build for us?

**Comment by danilo-silva-funttastic** [comment]
Hey @rapcmia,

we've been working here and I think we are good for a new test.
Can you please test our connector once again.

In order to help with the tests, here are some things we've observed:
- Sometimes the KUJI-DEMO market is working strangely on testnet, so it seems KUJI-USK might be better for this time.
- The statistics seems to be working well now (although it may take some seconds for the status of the orders to be computed and then the statistics to be ok).
- On Kujira, we have that filled tab for orders, but for accountability please check if the order is 100% filled there, HB statistics will only consider 100% filled orders (as informed above, HB may need some time to update the order status and then update the statistics).
- Sometimes (usually when stopping in a "bad" moment for HB) stopping and starting again could have an error like "order_books_initialized: False". Kujira, differently from Injective and Dexalot does not rely on orderbook snapshots but uses a chain python API to maintain it, because of that we've created a workaround that will dispatch the process to start the network again if it hasn't started. So stopping and starting again after a few seconds should work, as long as the RPC is working well.
- Sometimes we also receive an error (also usually when stopping in a "bad" moment for HB) "market is offline". It seems this happens when the dangling orders could not being canceled (maybe due to timeout or error when calling the RPC). It seems that calling stop again, and wait for the orders to be cancelled, make the market available again.

Please kind let us know if you find any further inconsistencies.

**Comment by danilo-silva-funttastic** [comment]
@fengtality, @rapcmia, @cardosofede 

We've tried the most we could do not touch the "base" or "core" classes, but we've found out that Injective and Dexalot is not relying completely on the Gateway, as we are. They are some APIs from their respective chains to maintain some processes there, especially regarding the orderbook.

Because of that we needed to touch the following (although we tried to make it to work the same way it was for the other connectors that does not have extra functionalities): hummingbot/connector/gateway/clob_spot/data_sources/clob_api_data_source_base.py
hummingbot/connector/gateway/clob_spot/gateway_clob_spot.py

Basically the main problem here is that when we start a strategy a orderbook tracker process is started (which fetches the orderbook snapshot) and when the strategy is stopped this process is killed. But the main problem is that when we start it again this orderbook tracker was not being started again (Injective and Dexalot were not being impacted because they don't use Gateway orderbook snapshots). We've then created a workaround, when the status check pass, if the start_network has not started, we start it, which starts the orderbook tracker too in the end.

We've also seen that because we don't have our extra routes we were being impacted compared to the other ones (like I explained they use extra functionality calling Python APIs not related to the Gateway) and because we tried to create an implementation using only the Gateway we suffered some difficulties. I'd say, we should probably have an space for extra routes for the connectors (although the main clob routes would be the same). One common route is to settle the funds (send the filled orders funds to the main wallet again), for example.

One other thing I'd like to comment is considering the current design of the CLOB routes, their design is not always consistent, mixing plural with singular, using strange endpoints (like the batch one which accepts orders to be created and cancelled at the same time) and some returns that are injective specific. I'd suggest to create a bounty to refactor these endpoints, especially now that we have only some few connectors there.

File: hummingbot/connector/gateway/clob_spot/data_sources/clob_api_data_source_base.py
Comment (cardosofede) [inline hummingbot/connector/gateway/clob_spot/data_sources/clob_api_data_source_base.py:48]: I think that this modifications are more specific to Kujira and you should inherit CLOBAPIDataSourceBase and implement your changes there

File: hummingbot/connector/gateway/clob_spot/data_sources/kujira/kujira_helpers.py
Comment (cardosofede) [inline hummingbot/connector/gateway/clob_spot/data_sources/kujira/kujira_helpers.py:38]: cool decorator!!

File: hummingbot/connector/gateway/clob_spot/data_sources/kujira/kujira_api_data_source.py
Comment (cardosofede) [inline hummingbot/connector/gateway/clob_spot/data_sources/kujira/kujira_api_data_source.py:121]: why we only support limit if in the constants we have Market too? will be good to have the two of them

**Review by cardosofede — CHANGES_REQUESTED** [review]

File: hummingbot/connector/gateway/clob_spot/data_sources/kujira/kujira_helpers.py
Comment (danilo-silva-funttastic) [reply hummingbot/connector/gateway/clob_spot/data_sources/kujira/kujira_helpers.py:38]: Hehe, thanks, it's really helpful ;)
I was thinking about creating a utils.py file so everyone could reuse it,
but then we preferred to not create extra files that would live outside of our connector for simplicity.

We have something similar on our Gateway implementation as well, but this one can cover the client too (which something is helpful to avoid disconnections or broken pipe errors).

File: hummingbot/connector/gateway/clob_spot/gateway_clob_spot.py
Comment (danilo-silva-funttastic) [reply hummingbot/connector/gateway/clob_spot/gateway_clob_spot.py:127]: Yeah, we really didn't want to touch this one here but the problem is the data source base classes (GatewayCLOBAPIDataSourceBase or CLOBAPIDataSourceBase) doesn't have this method available there and then we cannot call it. We used this technique that we've found somewhere in hummingbot before to enable the possibility of the method to be called if it exists.

But if it is ok for you, we could touch some more files and we could replicate this method on the CLOBAPIDataSourceBase and make it return False by default and then we could override it there for us (in our case returning True).

Because GatewayCLOBSpot and CLOBAPIDataSourceBase have different hierarchies some of the functionalities are not available in one another.

**Comment by danilo-silva-funttastic** [comment]
Hey @cardosofede, so, after your feedback we've changed our approach a little bit.
This time changing a little bit the default behavior of the classes but avoiding using that `hasattr` approach there.

As I commented there, the main problem is the fact the connect itself relies on the GatewayCLOBSpot class where the data sources are from another hierarchy (CLOBAPIDataSourceBase and GatewayCLOBAPIDataSourceBase). Because of that, if we need, from the datasource, override a behavior of the connector (like the way orders the orders are being cancelled, or the connector starting or stopping) we need to expose the connect instance to the data sources (what we called parent) and expose some methods for the data sources to override (most notably the start and stop ones).

Please kindly take a look at this new version when you have time :)

**Review by danilo-silva-funttastic — COMMENTED** [review]
Please take a look when you have time. We might be good to go with the current changes.

File: hummingbot/connector/gateway/clob_spot/gateway_clob_spot.py
Comment (cardosofede) [reply hummingbot/connector/gateway/clob_spot/gateway_clob_spot.py]: and how is possible that the cancel request is synchronous? don't you have to wait for a tx confirmation?

File: hummingbot/connector/gateway/clob_spot/data_sources/kujira/kujira_api_data_source.py
Comment (danilo-silva-funttastic) [reply hummingbot/connector/gateway/clob_spot/data_sources/kujira/kujira_api_data_source.py:26]: Nice catch! ;) We've updated it here.

File: hummingbot/connector/gateway/clob_spot/data_sources/kujira/kujira_api_data_source.py
Comment (danilo-silva-funttastic) [reply hummingbot/connector/gateway/clob_spot/data_sources/kujira/kujira_api_data_source.py:43]: Updated

File: hummingbot/connector/gateway/clob_spot/gateway_clob_spot.py
Comment (danilo-silva-funttastic) [reply hummingbot/connector/gateway/clob_spot/gateway_clob_spot.py]: We actually wait for the tx confirmation (the operation is async in term that we need to use an "await" to wait for it).

But in this case (at least from what we've understood), the GatewayCLOBSpot class implementation concerns more if the cancellation request was dispatched and later checked or if it is expected it to return the tx by the end of the execution of the method (which is our case).

File: hummingbot/connector/gateway/clob_spot/data_sources/kujira/kujira_api_data_source.py
Comment (danilo-silva-funttastic) [reply hummingbot/connector/gateway/clob_spot/data_sources/kujira/kujira_api_data_source.py:122]: I've added the support to market orders here as well now, and updated our enums to remove non-used types.
But we usually use only limit orders.

**Comment by danilo-silva-funttastic** [comment]
> @danilo-silva-funttastic hey Danilo, I think that now you are not using .parent at all. Actually can you point me to the place where we have this custom behavior in Kujira. Also a ready method can't start a new async task. You expect from that method to give you a boolean if it's ready or not.

Oh, yes, I forgot to mention, after the new changes the parent wasn't really needed, but still there are some behaviors that are not delegated to the datasource classes, so an approach similar to this one may still be needed in the future.

What is really solving the problem now is the change to the start method on the GatewayCLOBSpot, but because this method is unfortunately sync we need to dispatch tasks (we would prefer to use HB helper call_sync method, but it seems without using nest_asyncio package and applying it to the very beginning of HB (hummingbot.py / hummingbot_quickstart.py) it wouldn't work.

```
    def start(self, *args, **kwargs):
        super().start(**kwargs)
        safe_ensure_future(self.start_network())
        safe_ensure_future(self._api_data_source.start())
```

I moved the dispatch of the `start_network` task to the `start` method, which we've seen is also called when starting and also after stopping and starting again (it wasn't present on GatewaySOLCLOB before).

All of these changes are mainly related to the fact that when we do stop/start, start_network was not being called, the the orderbook tracker not initialized, and then ready was not ok (that's why we first intercepted the ready method). But now if this approach to the start method it's actually better. There are some hidden flows which is difficult to track because of the Cython code that was not being easy for us to find other places to intercept.

**Review by cardosofede — APPROVED** [review]
LGTM!

**Comment by cardosofede** [comment]
@rapcmia @nikspz the PR is ready to be reviewed by QA. If the code changes I need to review it again so my approval is partial but just to don't block you.

@danilo-silva-funttastic I think that you accomplished the task well now without modifying the core classes. If you have time, will be good if you can squash your commits.

**Comment by danilo-silva-funttastic** [comment]
> @rapcmia @nikspz the PR is ready to be reviewed by QA. If the code changes I need to review it again so my approval is partial but just to don't block you.
> 
> @danilo-silva-funttastic I think that you accomplished the task well now without modifying the core classes. If you have time, will be good if you can squash your commits.

Thanks a lot @cardosofede, I'll be doing this here and also updating the code with the lasted changes from the HB development branch.

**Comment by danilo-silva-funttastic** [comment]
@cardosofede @rapcmia @nikspz, we have already rebased our commits.
If it is preferable to have only one commit, please let me know.

**Review by danilo-silva-funttastic — COMMENTED** [review]
@cardosofede, as requested we've squashed our commits to only one.

**Review by rapcmia — APPROVED** [review]
LGTM

- Build this PR with https://github.com/hummingbot/gateway/pull/138
- Gateway connection successful
- Connect wallet successful using kepler
  - users must use the passphrase as pkey, if invalid will return `invalid Kujira mnemonic`
  - on setup, when asked for node url `none` should be ok.
  - `none` refers to connector will automatically look any on the list 
	   image
	      - Found on `conf/kujira.yml`
- Balance shows ok (should be ok on mainnet too)
- Setup PMM and observe order creation on https://fin.kujira.network/trade/kujira193dzcmy7lwuj4eda3zpwwt9ejal00xva0vawcvhgsyyp5cfh6jyq66wfrf
	- All tests done using testnet harpoon 4 and assuming the same behavior for mainnet 
  - Order created and cancelled successfully
  - There are noticeable delays but no stuck orders 
  - Ran for couple of hours and observe behavior 
  - Filled order data between CSV and Client ok 
  - Txns are available and successful
- Build docker image and use `hummingbot_gateway_compose` tests
  - Setup other dex connectors to check if affected `{uniswap-arbitrum, uniswap-polygon}`, all ok 
  - gateway list shows `Kujira` 

---

Ran tests base on bronze standard and community is advised that this connector should work ok also on mainnet. Should there be issues, community should create ticket so the developers will be inform for fix or possible improvements when used for long term scenarios. Thank you

**Comment by cardosofede** [comment]
@danilo-silva-funttastic I'm checking the code and you reverted the changes that we discussed before

**Comment by danilo-silva-funttastic** [comment]
> @danilo-silva-funttastic I'm checking the code and you reverted the changes that we discussed before

Hey @cardosofede, I was checking here but I haven't got what you mean.
The only changes we've made was to keep the branch up to date without conflicts.

I've also double checked the changes and they seem good to me (we are not using that parent mechanism as we used before, neither dispatching a async task from the ready method (since we didn't know about the start_network and stop_network method before)). There are some other few changes to correctly start and stop the network and to enable the possibility to use controle the timeout of the orders cancelation to avoid problems.

Please let me know if I missed something.

**Comment by cardosofede** [comment]
@danilo-silva-funttastic you are right my friend! I don't know why but when I opened the tab I saw the parent workflow.

**Review by cardosofede — APPROVED** [review]
LGTM!

**Comment by danilo-silva-funttastic** [comment]
> @danilo-silva-funttastic you are right my friend! I don't know why but when I opened the tab I saw the parent workflow.

Hehe, I see, thanks a lot @cardosofede my friend for confirming it's all good :)

By the way, before rebasing the branch today, I've seen the build wasn't passing because it didn't reach 80% coverage, I think this wasn't the case before, is this requirement relatively new? Previously we've seen that some of the unit tests Injective and Dexalot was quite the same (for example this one here: test/hummingbot/connector/gateway/clob_spot/data_sources/injective/test_injective_api_data_source.py). They only do some asserts on top of the mock responses of the gateway. Do you know if this would be mandatory for the merge or if we are good to go without these kind of tests?

**Comment by danilo-silva-funttastic** [comment]
@cardosofede, @fengtality, @rapcmia,

we've included a new file with test cases to cover our code, it's now in 80+%, so probably our build might be able to pass and be green now.

Can you please approve our workflow to run?

**Comment by cardosofede** [comment]
@danilo-silva-funttastic approved

**Comment by danilo-silva-funttastic** [comment]
> @danilo-silva-funttastic approved

Thanks a lot @cardosofede, it seems we needed to include a "__init__.py" file inside of the test folder, without it our test was not being discovered by the full suite. Can you please allow our build to run again? :)

Strangely enough the full suite seems to only work on Linux, at least for me, when using macOS, it doesn't run until the end.

**Comment by danilo-silva-funttastic** [comment]
> > @danilo-silva-funttastic approved
> 
> Thanks a lot @cardosofede, it seems we needed to include a "**init**.py" file inside of the test folder, without it our test was not being discovered by the full suite. Can you please allow our build to run again? :)
> 
> Strangely enough the full suite seems to only work on Linux, at least for me, when using macOS, it doesn't run until the end.


@cardosofede , sorry to bother once more, can you please allow our build to run once more?

Two of our tests were failing on the full workflow build, but now it should be all good ;)

I had an idea here and created a PR using our own fork and tested there, we now have a green build there and hopefully here as well

https://github.com/Team-Kujira/hummingbot/pull/2

**Comment by danilo-silva-funttastic** [comment]
> Hi @danilo-silva-funttastic, can you also open a PR to create the docs on our site for the release this month - https://github.com/hummingbot/hummingbot-site
> 
> You can use any of the bronze connectors docs as a template.

Hey @david-hummingbot, thanks for that, sure, we will be creating a PR there as well ;)

**Pull request merged**
